File: keylib.py
from bytelib import bytes_to_long,long_to_bytes

import base64
import struct

def decode(b):
    #return base64.b64decode(b).decode()
    return base64.b64encode(b).decode()
def encode(s):
    try:
        return base64.b64decode(s.encode())
    except:
        import logging
        logging.error(20,s)
        raise

class DerpKey(bytes):
    type = 'CHK'
    def __str__(self):
        return decode(self)
    def __repr__(self):
        return self.type+'('+decode(self)[:4]+')'

def Key(b,type='CHK'):
    if isinstance(b,str):
        b = encode(b)
    key = DerpKey(b)
    key.type = type
    return key

class DerpURI(DerpKey):
    depth = None
    def __repr__(self):
        return self.type+'URI('+decode(self)[:4]+','+str(self.depth)+')'

def URI(s,depth=None,type='CHK'):
    if depth and isinstance(s,bytes):
        data = struct.pack('B',depth)+s
        uri = DerpURI(data)
        uri.depth = depth
    elif len(data) == keySize:
        data = encode(s)
        uri = DerpURI(data)
        uri.depth = 1
    else:
        data = encode(s)
        uri = DerpURI(data[1:])
        uri.depth = data[0]
    uri.type = 'URI'+type
    return uri

def join(keys):
    return Key(b''.join(keys))

from hashlib import sha512

def makeHash(b):
    derp = sha512()
    derp.update(b)
    return Key(derp.digest())

keySize = len(makeHash(b''))

File: keylib.lua
local bytelib = require('bytelib')
local rockrequire = require('rockrequire')
local sodium = require('sodium')
local b2l = bytelib.bytes_to_long
local l2b = bytelib.long_to_bytes

local mime = require('mime') -- yuck

function lopbyte(data) 
    local byte = string.byte(data,1)
    return byte,string.sub(data,2)
end

function decode(data) 
    if type(data) ~= 'string' then return data end
    data = data.unb64(data)
    if string.len(data) == keySize then
        return Key(data)
    elseif string.len(data) == keySize+1 then
        local depth
        depth,data = lopbyte(data)
        return URI(data,depth)
    end
end

function Key(data)
    return {
        data = data,
        encode = function()
            return mime.b64(data)
        end,
        string = function()
            return 'Key('..mime.b64(data)..')'
        end
    }
end

function override(what,name,over) 
    local old = what[name]
    what[name] = function(...) return over(old,...) end
end

function URI(data, depth)
    if depth then
        key = Key(data)
        data = string.char(depth)..data
    else
        key = Key(string.sub(data,1))
        depth = string.byte(data,0)
    end

    return {
        key = key,
        encode = function()
            return mime.b64(data)
        end,
        string = function()
            return 'URI('..key.string()..','..depth..')'
        end,
        depth = depth
    }
end

function join(keys) 
    data = ""
    for i,key in ipairs(keys) do
        data = data .. key.data
    end
    return {
        data = data,
        encode = function() return mime.b64(data) end,
        string = function()
            return "KeyList("..#keys..")"
        end
    }
end

function split(data)
    local which = -1
    return function()
        which = which + 1
        local amt = which*sodium.hash.size
        return which,string.sub(data,amt,amt+sodium.hash.size)
    end
end

function makeHash(b)
    return sodium.hash(b)
end

local keySize = sodium.hash.size

return {
    URI = URI,
    Key = Key,
    decode = decode,
    join = join,
    split = split,
    keySize = keySize
}
